Src/VsVimShared/VsVimHost.cs

```diff
@@ -456,6 +456,58 @@ public override void EndBulkOperation()
             }
         }
 
+        /// <summary>
+        /// Perform the 'find in files' operation using the specified parameters
+        /// </summary>
+        /// <param name="pattern">BCL regular expression pattern</param>
+        /// <param name="matchCase">whether to match case</param>
+        /// <param name="filesOfType">which files to search</param>
+        /// <param name="flags">flags controlling the find operation</param>
+        public override void FindInFiles(string pattern, bool matchCase, string filesOfType, VimGrepFlags flags)
+        {
+            // Callback to navigate to the first error.
+            void onFindDone(vsFindResult result, bool cancelled)
+            {
+                _dte.Events.FindEvents.FindDone -= onFindDone;
+
+                if (!cancelled && result == vsFindResult.vsFindResultFound)
+                {
+                    SafeExecuteCommand(null, "Edit.GoToFindResults1NextLocation");
+                }
+            }
+
+            try
+            {
+                // Because 'find all' is asynchronous, we need to use an event
+                // to support 'jump to first'. However, this doesn't appear to
+                // work reliably.
+                var jumpToFirst = !flags.HasFlag(VimGrepFlags.NoJumpToFirst);
+                jumpToFirst = false;
+
+                if (_dte.Find is EnvDTE80.Find2 find)
+                {
+                    find.Action = vsFindAction.vsFindActionFindAll;
+                    find.FindWhat = pattern;
+                    find.Target = vsFindTarget.vsFindTargetSolution;
+                    find.MatchCase = matchCase;
+                    find.MatchWholeWord = false;
+                    find.PatternSyntax = vsFindPatternSyntax.vsFindPatternSyntaxRegExpr;
+                    find.FilesOfType = filesOfType;
+                    find.ResultsLocation = vsFindResultsLocation.vsFindResults1;
+                    find.WaitForFindToComplete = false;
+                    if (jumpToFirst)
+                    {
+                        _dte.Events.FindEvents.FindDone += onFindDone;

```

Comment:
In case you haven't run into this problem before, the various event types in DTE can have non-intuitive garbage collection behavior. Because they're created on access, and not stored as a reference in the containing type, they can get garbage collected before the events are raised. Example is `SolutionEvents`

[https://docs.microsoft.com/en-us/dotnet/api/envdte.solutionevents?view=visualstudiosdk-2017#remarks](https://docs.microsoft.com/en-us/dotnet/api/envdte.solutionevents?view=visualstudiosdk-2017#remarks)

As far as I can tell `FindEvents` doesn't suffer from this problem. At the same time this could just be an omission from the documentation. If this is working for you though then likely the documentation is correct here.

Reply:
Ha, ha, I **did** run into that problem: the 'on find done' callback only worked about 50% of the time. I looked into it without solving it and eventually gave up. I attributed the problem to the fact that `FindDone` was marked "Microsoft Internal Use Only". For that matter, so are all the `SolutionEvents`. I would have eventually figured it out if the `FindDone` documentation contained a similar remarks section.

In any case, I am very ambivalent about the "jump to first" feature. I am not sure I will ever want to use it, and yet it is the default. Let me get it working reliably and I will test it out. Thanks for solving the mystery because it was bugging me.